src/seq_mv/csr_matop.c

```diff
@@ -373,29 +400,78 @@ hypre_CSRMatrixMultiplyHost( hypre_CSRMatrix *A,
       }
 
       B_marker = hypre_CTAlloc(HYPRE_Int, ncols_B, HYPRE_MEMORY_HOST);
-
       for (ib = 0; ib < ncols_B; ib++)
+      {
          B_marker[ib] = -1;
+      }
 
+      /* First pass: compute sizes of C rows. */
       num_nonzeros = 0;
-      for (ic = ns; ic < ne; ic++)
+      if (rownnz_A == NULL)
       {
-         C_i[ic] = num_nonzeros;
-         if (allsquare)
+	 if (allsquare)
```

Comment:
Why are you reorganizing this?
The case of a square matrix only needs different treatment of the diagonal element. So, it's a small preamble before the rest. This seems to be a lot of extra code.

Reply:
I followed the same reasoning as for rownnz: "moving the conditional statement outside the for loop could improve performance". But I realized last week that since it's a single branch which is evaluated O(10^5) times (estimate valid for a 100.000 dofs problem), it actually does not improve performance. Maybe it's better to revert this for better code maintenance. Should I do this for `rownnz` as well?